Tpm2 specify pcr value

ddstreet opened the pull request:
Allow specifying specific PCR values to systemd-cryptenroll. This allows using expected (calculated/measured) values instead of only being able to enroll using the current tpm pcr values.

<!-- devel-freezer = {"comment-id":"1635970769","freezing-tag":"v254-rc3"} -->

ddstreet commented:
This has the same goal as #28398

ddstreet commented:
I wanted to get this PR opened asap, but I don't think this needs to get in until after the release.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -4127,6 +4127,30 @@ int tpm2_parse_luks2_json(
+        return 0;
+}
+
+int tpm2_hash_alg_from_size(size_t size) {
```
uh, this smells like it's going to break sooner or later, no? if another hash alg is added with the same size?

ddstreet replied:
quite true, and that's [already the case](https://en.wikipedia.org/wiki/SHA-2#Comparison_of_SHA_functions) for sha3 family of hashes (which have the same size as sha2 hashes). I'll rework this.

I added this in order to avoid users needing to specify the hash to --tpm2-pcrs, since all the sha1/sha2 algs have different hash sizes. Do you think we should always require the hash alg, or should we keep the 'detection' of the right hash but only for the sha1/sha2 algs?

for example should we allow this:
`systemd-cryptenroll --tpm2-pcrs 1=0x0000000000000000000000000000000000000000000000000000000000000000`
which would 'detect' that the hash is sha256 from the hash length

while if users wanted to use sha3-256, they could specify the alg (technically this won't work now, since we don't parse for sha3 yet):
`systemd-cryptenroll --tpm2-pcrs 1:sha3-256=0x0000000000000000000000000000000000000000000000000000000000000000`

Making the hash alg optional is strictly for the convenience of users, so if you think we should require it i have no objections.

ddstreet replied:
i pushed a new version with `tpm2_hash_alg_from_size()` removed, and `tpm2_parse_pcr_digest_spec()` doing the SHA1/SHA2 'detection' based on hash value length. I can update to remove that and require specifying the hash alg if you think that's better, let me know.

ddstreet replied:
after looking at it again, i removed the 'optional' part - a hash algorithm  has to be provided now when specifying a hash digest value. the 'convenience' of omitting the hash alg isn't really needed.

ddstreet replied:
i removed the CLEANUP_ERASE() in favor of TAKE_STRUCT() and re-pushed...figured we dont need to do both.

ddstreet commented:
changing to draft until the next release.

DaanDeMeyer commented:
> changing to draft until the next release.

No need, we have a special label for stuff that should be merged after the next release.

ddstreet commented:
> > changing to draft until the next release.
> 
> No need, we have a special label for stuff that should be merged after the next release.

ah ok, unmarking it as draft then, thanks

bluca commented on src/shared/tpm2-util.c:
```diff
@@ -3678,104 +3678,28 @@ int tpm2_tpm2b_public_from_pem(const void *pem, size_t pem_size, TPM2B_PUBLIC *r
-        assert(TPM2_PCR_MASK_VALID(hash_pcr_mask));
-        assert(TPM2_PCR_MASK_VALID(pubkey_pcr_mask));
-
-        /* So here's what we do here: we connect to the TPM2 chip. It persistently contains a "seed" key that
```
No need to remove this comment?

ddstreet replied:
ack, updated to leave it

bluca reviewed (changes requested):
Some minor comments

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -4187,6 +4187,18 @@ int tpm2_parse_luks2_json(
+                return 48;
+        if (alg == TPM2_ALG_SHA512)
+                return 64;
+        return 0;
```
isn't asserting a bit too harsh? The other hash alg to * functions similarly return NULL if the alg isn't in the list. Maybe a debug log msg would be better than assert?

bluca replied:
depends - is it a programming error, or a user input error that would get here? if the former an assert is right, if the latter then yes a log message is enough

ddstreet replied:
I added debug logs to each of the alg to/from string (and size) functions

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1885,22 +2199,27 @@ static int tpm2_create_loaded(
+/* Read hash values from the specified PCR selection. Provides a Tpm2PCRValue array that contains all
+ * requested PCR values, in the order provided by the TPM. Normally, the provided pcr values will match
+ * exactly what is in the provided selection, but the TPM may ignore some selected PCRs (for example, if an
+ * unimplemented PCR index is requested), in which case those PCRs will be absent from the provdied pcr
```
fixed, thanks

bluca commented on src/resolve/resolved-dns-dnssec.c:
```diff
@@ -14,7 +14,7 @@
 #include "string-table.h"
 
-#if PREFER_OPENSSL
+#if PREFER_OPENSSL && OPENSSL_VERSION_MAJOR >= 3
```
got it

ddstreet commented on src/shared/openssl-util.c:
```diff
@@ -128,6 +128,148 @@ int rsa_pkey_to_suitable_key_size(
+        if (!RSA_set0_key(rsa_key, bn_n, bn_e, NULL))
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set RSA n/e.");
+        /* rsa_key owns these now, don't free */
+        bn_n = bn_e = NULL;
```
do you mean use `TAKE_PTR()` directly as the param, e.g.
`RSA_set0_key(rsa_key, TAKE_PTR(bn_n), TAKE_PTR(bn_e), NULL)`

I thought about doing that and I would certainly prefer it, but my concern is about leaking them if the called function returns error...

ddstreet replied:
got it, thanks - updated

ddstreet commented on src/shared/openssl-util.c:
```diff
@@ -128,6 +128,148 @@ int rsa_pkey_to_suitable_key_size(
+
+        _cleanup_(BN_freep) BIGNUM *bn_n = BN_bin2bn(n, n_size, NULL);
+        if (!bn_n)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to create BIGNUM for RSA n.");
```
It does looks like it provides some functionality to check for error details:
https://www.openssl.org/docs/man3.0/man3/ERR_get_error.html

I haven't tried using it to see how informative the errors are (or if it even works for all errors). I could look into it as it certainly would be better than the generic messages.

ddstreet replied:
I'll look at this for a separate PR if that's ok

bluca replied:
sure

ddstreet commented on src/shared/openssl-util.c:
```diff
@@ -270,6 +270,199 @@ int rsa_pkey_new(size_t bits, EVP_PKEY **ret) {
+        if (!EVP_PKEY_assign_EC_KEY(pkey, eckey))
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to assign ECC key.");
+        /* pkey owns this now, don't free */
+        eckey = NULL;
```
i updated this (and the earlier ones) to use TAKE_PTR() after the function call, in case that's what you meant.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -3593,6 +3474,210 @@ static int tpm2_build_sealing_policy(
+                                point->y.buffer,
+                                point->y.size,
+                                ret);
+        } else if (p->type == TPM2_ALG_RSA) {
```
ack, updated

yuwata commented on src/shared/tpm2-util.c:
```diff
@@ -1237,16 +1237,40 @@ void tpm2_tpms_pcr_selection_from_mask(uint32_t mask, TPMI_ALG_HASH hash_alg, TP
 }
 
+/* Test if all bits in the mask are set in the TPMS_PCR_SELECTION. */
+bool tpm2_tpms_pcr_selection_has_mask(const TPMS_PCR_SELECTION *s, uint32_t mask) {
+        uint32_t m;
```
Missing assertion for `s`.

ddstreet replied:
ack, updated

yuwata commented on src/shared/tpm2-util.c:
```diff
@@ -1443,6 +1466,34 @@ void tpm2_tpml_pcr_selection_sub_tpms_pcr_selection(TPML_PCR_SELECTION *l, const
 }
 
+/* Test if all bits in the mask for the hash are set in the TPML_PCR_SELECTION. */
+bool tpm2_tpml_pcr_selection_has_mask(const TPML_PCR_SELECTION *l, TPMI_ALG_HASH hash, uint32_t mask) {
+        TPMS_PCR_SELECTION *s = tpm2_tpml_pcr_selection_get_tpms_pcr_selection((TPML_PCR_SELECTION*) l, hash);
```
missing assertion for `l`.

Also, is it not necessary to copy `l`, like `tpm2_tpml_pcr_selection_to_mask()` ?
Or, maybe, it is better to reuse `tpm2_tpml_pcr_selection_to_mask()` ?

Anyway, the cast sounds spurious and dangerous.

ddstreet replied:
yep you are correct, i'll update this to use `tpm2_tpml_pcr_selection_to_mask()` instead

yuwata commented on src/shared/tpm2-util.c:
```diff
@@ -1542,6 +1543,311 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+        assert(a);
+        assert(b);
+
+        return ((int) a->hash) - ((int) b->hash) ?: ((int) a->index) - ((int) b->index);
```
Please use `CMP()` macro.

ddstreet replied:
ah thanks i had meant to go back and fix this, will update it

ddstreet commented:
Updated per review comments, and I also added commit [daa226a](https://github.com/systemd/systemd/pull/28398/commits/daa226a76070a6f7304899ada29d2da1e5b473f8) to change the *_to_mask() functions to return the mask, which I think is easier to use and makes more sense.

yuwata commented:
Sorry, please rebase again.

bluca commented:
@ddstreet this needs a rebase on main, there are conflicts, then we can merge it

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3635,6 +3516,212 @@ static int tpm2_build_sealing_policy(
+        case NID_X9_62_prime256v1: *ret = TPM2_ECC_NIST_P256; return 0;
+        case NID_secp384r1:        *ret = TPM2_ECC_NIST_P384; return 0;
+        case NID_secp521r1:        *ret = TPM2_ECC_NIST_P521; return 0;
+        case NID_sm2:              *ret = TPM2_ECC_SM2_P256;  return 0;
```
that's a strange way of indentation, how we never do it otherwise.

maybe encode this in a static const array? 

```c
static const struct {
        int openssl_curve_id;
        TPM2_ECC_CURVE tpm2_ecc_curve;
} table[] = {
        { NID_X9_62_prime192v1, TPM2_ECC_NIST_P192 },
        …
};

FOREACH_ARRAY(t, table, ELEMENTSOF(table)) 
        if (t->openssl_curve_id == curve_id)
                 return t->tpm2_ecc_curve;
…
```

ddstreet replied:
ack

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3635,6 +3516,212 @@ static int tpm2_build_sealing_policy(
+        case TPM2_ECC_NIST_P256: *ret = NID_X9_62_prime256v1; return 0;
+        case TPM2_ECC_NIST_P384: *ret = NID_secp384r1;        return 0;
+        case TPM2_ECC_NIST_P521: *ret = NID_secp521r1;        return 0;
+        case TPM2_ECC_SM2_P256:  *ret = NID_sm2;              return 0;
```
similar here

ddstreet replied:
ack

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3635,6 +3516,212 @@ static int tpm2_build_sealing_policy(
+        assert(ret);
+
+        const TPMT_PUBLIC *p = &public->publicArea;
+        if (p->type == TPM2_ALG_ECC) {
```
sounds like a candidate for switch()?

ddstreet replied:
ack

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3635,6 +3516,212 @@ static int tpm2_build_sealing_policy(
+        key_id = EVP_PKEY_id(pkey);
+#endif
+
+        if (key_id == EVP_PKEY_EC) {
```
switch?

ddstreet replied:
ack

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3635,6 +3516,212 @@ static int tpm2_build_sealing_policy(
+                TPM2_ECC_CURVE curve;
+                r = tpm2_ecc_curve_from_openssl_curve_id(curve_id, &curve);
+                if (r < 0)
+                        return r;
```
on this error you debug log, but on the error one earlier you log at LOG_ERR log level. please make sure this function either logs about all errors at LOG_ERR or about none.

(given this is more "library-like" and less "main-function-like", it should probably stick to debug logging, and let callers log at a higher log level)

ddstreet replied:
ack

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3635,6 +3516,212 @@ static int tpm2_build_sealing_policy(
+                                               "RSA key e size %zu too large.", e_size);
+
+                uint32_t exponent = 0;
+                memcpy((void*) &exponent, e, e_size);
```
the void cast should be unnecessary?

ddstreet replied:
ack

poettering commented on src/cryptenroll/cryptenroll-tpm2.c:
```diff
@@ -206,23 +206,69 @@ int enroll_tpm2(struct crypt_device *cd,
+                if (hash_pcr_values[i].value.size > 0) {
+                        pcr_value_specified = true;
+                        break;
+                }
```
mabe make this loop a helper function of its own?

ddstreet replied:
ack, added `tpm2_pcr_values_has_any_values()` and `tpm2_pcr_values_has_all_values()`

poettering commented on src/shared/tpm2-util.h:
```diff
@@ -240,7 +241,8 @@ int tpm2_make_luks2_json(int keyslot, uint32_t hash_pcr_mask, uint16_t pcr_bank,
 
 /* Default to PCR 7 only */
-#define TPM2_PCR_MASK_DEFAULT (UINT32_C(1) << 7)
+#define TPM2_PCR_INDEX_DEFAULT (7)
```
should probably be 7U or UINT32_C(7)

ddstreet replied:
ack

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -391,13 +391,22 @@
         specifies that PCR registers 4, 1, and 5 should be used.</para>
+        <para>Example: <option>--tpm2-pcrs=7:sha256</option> specifies that PCR register 7 from the SHA256
+        bank should be used.</para>
+        <para>Example: <option>--tpm2-pcrs=4:sha1=0x3A3F780F11A4B49969FCAA80CD6E3957C33B2275</option>
```
hmm, all our tools output PCR values in lower case and without the 0x prefix. given the prefix is redundant, can you remove it, and lower case it?

ddstreet replied:
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -4239,6 +4239,18 @@ int tpm2_parse_luks2_json(
+        if (alg == TPM2_ALG_SHA384)
+                return 48;
+        if (alg == TPM2_ALG_SHA512)
+                return 64;
```
ack, will change the `tpm2_*_alg_to_*()` functions to use switch()

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
         return weight;
 }
 
+bool TPM2_PCR_VALUE_VALID(const Tpm2PCRValue *pcr_value) {
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+                if (r < 0)
+                        return false;
+
+                if ((int) pcr_value->value.size != r) {
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+ * 1) all entries must be sorted in ascending order (e.g. using tpm2_sort_pcr_values())
+ * 2) all entries must be unique, i.e. there cannot be 2 entries with the same hash and index
+ */
+bool TPM2_PCR_VALUES_VALID(const Tpm2PCRValue *pcr_values, size_t n_pcr_values) {
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+bool TPM2_PCR_VALUE_VALID(const Tpm2PCRValue *pcr_value) {
+        int r;
+
+        assert(pcr_value);
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+bool TPM2_PCR_VALUES_VALID(const Tpm2PCRValue *pcr_values, size_t n_pcr_values) {
+        assert(pcr_values || n_pcr_values == 0);
+
+        for (size_t i = 0; i < n_pcr_values; i++) {
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid PCR values.");
+
+        for (size_t i = 0; i < n_pcr_values; i++)
+                if (pcr_values[i].hash == hash)
```
ack, updated all the appropriate `for() `in tpm2-util.c to use `FOREACH_ARRAY()`

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+
+                _cleanup_free_ void *buf = NULL;
+                size_t buf_size = 0;
+                r = unhexmem(p, strlen(p), &buf, &buf_size);
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+                        return log_error_errno(r, "Invalid pcr hash value '%s': %m", p);
+
+                pcr_value.value.size = buf_size;
+                assert(sizeof(pcr_value.value.buffer) >= pcr_value.value.size);
```
i think this was addressed in the final merged code

ddstreet replied:
by using TPM2B_DIGEST_CHECK_SIZE()

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+                pcr_value.hash = (TPMI_ALG_HASH) r;
+        }
+
+        if (!isempty(p)) {
```
ack, updated to nest the second block.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+        int r;
+
+        r = asprintf(&index, "%u", pcr_value->index);
+        if (r < 0)
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+        if (r < 0)
+                return NULL;
+
+        const char *hash = tpm2_hash_alg_to_string(pcr_value->hash);
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+                        return NULL;
+        }
+
+        return strjoin(index, hash ? ":" : "", hash ?: "", value ? "=" : "", value ?: "");
```
ack

ddstreet commented on src/shared/tpm2-util.h:
```diff
@@ -88,6 +88,26 @@ int tpm2_handle_new(Tpm2Context *context, Tpm2Handle **ret_handle);
+        TPM2B_DIGEST value;
+} Tpm2PCRValue;
+
+#define TPM2_PCR_VALUE_MAKE(i, h, v) (Tpm2PCRValue) { .index = (i), .hash = (h), .value = ((TPM2B_DIGEST) v), }
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1520,6 +1521,318 @@ size_t tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+                if (!GREEDY_REALLOC_APPEND(pcr_values, n_pcr_values, &pcr_value, 1))
+                        return log_oom();
+        }
+
```
since the `tpm2_parse_pcr_argument()` function does both sorting *and* default-hash application, the `tpm2_pcr_values_from_string()` function can't also do sorting. For example:
`7:sha256,4:sha1,0'
should, according to the function doc, use the sha256 bank for index 0. But if sorting was done before default-hash application, index 0 would be assigned sha1.

An obvious second argument is, should we *ever* accept PCRs from multiple banks? And the answer there could be, no, in which case the sorting could be done directly in `tpm2_pcr_values_from_string()`, and `tpm2_parse_pcr_argument()` could just return error if multiple banks are specified. I went with the more flexible code, but I don't have strong feelings about locking it down to a single bank.

In either cawe, since any user-provided pcr values should be handled by `tpm2_parse_pcr_argument()` instead of being directly processed by `tpm2_pcr_values_from_string()`, i don't think end users will need to do any sorting themselves.

ddstreet commented on src/test/test-tpm2.c:
```diff
@@ -619,50 +619,58 @@ TEST(calculate_policy_authorize) {
-        digest_init_sha256(&pcr_values[14], "e3991b7ddd47be7e92726a832d6874c5349b52b789fa0db8b558c69fea29574e");
-        digest_init_sha256(&pcr_values[15], "852dae3ecb992bdeb13d6002fefeeffdd90feca8b378d56681ef2c885d0e5137");
-        assert_se(tpm2_calculate_policy_pcr(&pcr_selection, pcr_values, 16, &d) == 0);
+        Tpm2PCRValue v2[] = {
```
ack

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -4678,29 +4639,147 @@ Tpm2Support tpm2_support(void) {
+                for (size_t i = 0; i < n_pcr_values; i++)
+                        if (pcr_values[i].hash == 0)
+                                pcr_values[i].hash = default_hash;
+}
```
no, `tpm2_pcr_values_from_string()` calls `tpm2_pcr_value_from_string()` for each individual pcr value, and for that function the hash and value are optional. So the hash can be 0 here.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -4678,29 +4639,147 @@ Tpm2Support tpm2_support(void) {
+
+        tpm2_pcr_values_apply_default_hash_alg(pcr_values, n_pcr_values);
+
+        tpm2_sort_pcr_values(pcr_values, n_pcr_values);
```
no, for the default-hash reasoning I gave earlier.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -4678,29 +4639,147 @@ Tpm2Support tpm2_support(void) {
+                return r;
+
+        /* If we got previous values, append them. */
+        if (*ret_pcr_values && !GREEDY_REALLOC_APPEND(pcr_values, n_pcr_values, *ret_pcr_values, *ret_n_pcr_values))
```
ack

ddstreet commented on src/shared/tpm2-util.h:
```diff
@@ -153,6 +153,62 @@ int tpm2_calculate_policy_pcr(const Tpm2PCRValue *pcr_values, size_t n_pcr_value
+                struct_type UNIQ_T(STRUCT, uniq) = { .size_field = UNIQ_T(SIZE, uniq), }; \
+                assert(sizeof(UNIQ_T(STRUCT, uniq).buffer_field) >= (size_t) UNIQ_T(SIZE, uniq)); \
+                if (UNIQ_T(BUF, uniq))                                  \
+                        memcpy(UNIQ_T(STRUCT, uniq).buffer_field, UNIQ_T(BUF, uniq), UNIQ_T(SIZE, uniq)); \
```
ack

ddstreet commented on src/shared/openssl-util.c:
```diff
@@ -271,6 +271,199 @@ int rsa_pkey_new(size_t bits, EVP_PKEY **ret) {
+        if (!EVP_PKEY_get_utf8_string_param(pkey, OSSL_PKEY_PARAM_GROUP_NAME, NULL, 0, &name_size))
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to get ECC group name size.");
+
+        _cleanup_free_ char *name = malloc(name_size + 1);
```
ack

ddstreet commented on src/test/test-tpm2.c:
```diff
@@ -619,50 +619,58 @@ TEST(calculate_policy_authorize) {
-        digest_init_sha256(&pcr_values[1], "aa1154c9e0a774854ccbed4c8ce7e9b906b3d700a1a8db1772d0341a62dbe51b");
-        digest_init_sha256(&pcr_values[2], "cfde439a2c06af3479ca6bdc60429b90553d65300c5cfcc40004a08c6b5ad81a");
-        assert_se(tpm2_calculate_policy_pcr(&pcr_selection, pcr_values, 3, &d) == 0);
+        Tpm2PCRValue v1[] = {
```
actually, if we static const these vars, we have to cast from TPM2_PCR_VALUE_MAKE() because that isn't const (and can't be used for a static initializer), and it really doesn't feel like it's worth it for test code. I'm not sure it would really even improve anything.

I'm going to leave these local vars, but if you feel strongly about it I can update them.
